### PR TITLE
resharding on new nodes [ready for review]

### PR DESCRIPTION
**Pull request by rlan35**
Tested locally with 4 shards * 5 nodes and 5 new nodes.

New nodes can successfully issue staking transaction and be added into one of the shard after resharding.

New nodes once receives the resharding message can also automatically restart to join the assigned shard and participate in consensus there.

### CODE REVIEW

File: cmd/harmony/main.go [hunk 1]
```diff
@@ -72,7 +73,7 @@ var (
 	ip               = flag.String("ip", "127.0.0.1", "IP of the node")
 	port             = flag.String("port", "9000", "port of the node.")
 	logFolder        = flag.String("log_folder", "latest", "the folder collecting the logs of this execution")
-	freshDB          = flag.Bool("fresh_db", false, "true means the existing disk based db will be removed")
+	freshDB          = flag.Bool("fresh_db", true, "true means the existing disk based db will be removed")
```
Comment (LeoHChen) [inline cmd/harmony/main.go]: why change the default value?

File: core/resharding.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ const (
 	// GenesisShardNum is the number of shard at genesis
 	GenesisShardNum = 4
 	// GenesisShardSize is the size of each shard at genesis
-	GenesisShardSize = 50
+	GenesisShardSize = 5
```
Comment (LeoHChen) [inline core/resharding.go]: don't change the default value.

File: core/resharding.go [hunk 3]
```diff
@@ -94,25 +94,35 @@ func (ss *ShardingState) cuckooResharding(percent float64) {
...
+	rand.Seed(int64(ss.rnd))
+	ss.sortCommitteeBySize()
+
+	// Take out and preserve leaders
+	leaders := []types.NodeID{}
 	for i := 0; i < ss.numShards; i++ {
-		// At genesis epoch, the shards are empty.
 		if len(ss.shardState[i].NodeList) > 0 {
+			leaders = append(leaders, ss.shardState[i].NodeList[0])
+			utils.GetLogInstance().Info("LEADER===01", "shard", ss.shardState[i])
```
Comment (LeoHChen) [inline core/resharding.go]: please remove the debug log.

File: core/resharding.go [hunk 4]
```diff
@@ -94,25 +94,35 @@ func (ss *ShardingState) cuckooResharding(percent float64) {
...
 	}
-}
 
-// Reshard will first add new nodes into shards, then use cuckoo rule to reshard to get new shard state
-func (ss *ShardingState) Reshard(newNodeList []types.NodeID, percent float64) {
-	rand.Seed(int64(ss.rnd))
-	ss.sortCommitteeBySize()
-	// TODO: separate shuffling and leader assignment
-	ss.assignLeaders()
+	utils.GetLogInstance().Info("LEADER===1", "leaders", leaders)
```
Comment (LeoHChen) [inline core/resharding.go]: please remove the debug log.

File: core/resharding.go [hunk 5]
```diff
@@ -94,25 +94,35 @@ func (ss *ShardingState) cuckooResharding(percent float64) {
...
 	ss.assignNewNodes(newNodeList)
 	ss.cuckooResharding(percent)
+
+	// Put leader back
+	for i := 0; i < ss.numShards; i++ {
+		ss.shardState[i].NodeList = append(leaders[i:i+1], ss.shardState[i].NodeList...)
```
Comment (LeoHChen) [inline core/resharding.go]: please do index out of range check here.

File: core/resharding.go [hunk 6]
```diff
@@ -94,25 +94,35 @@ func (ss *ShardingState) cuckooResharding(percent float64) {
...
+		ss.shardState[i].Leader = leaders[i]
+		utils.GetLogInstance().Info("LEADER===2", "shard", ss.shardState[i])
```
Comment (LeoHChen) [inline core/resharding.go]: debug log.

File: node/node.go [hunk 7]
```diff
@@ -427,6 +438,8 @@ func (node *Node) initNodeConfiguration() (service.NodeConfig, chan p2p.Peer) {
 func (node *Node) AddBeaconChainDatabase(db ethdb.Database) {
 	database := db
 	if database == nil {
+		fmt.Println("CHAIN ID===NEW DB")
```
Comment (LeoHChen) [inline node/node.go]: debug log please remove

File: node/node_handler.go [hunk 8]
```diff
@@ -601,23 +628,43 @@ func (node *Node) processEpochShardState(epochShardState *types.EpochShardState)
...
 			}
-
-			utils.GetLogInstance().Info("Restarting program", "args", os.Args, "env", os.Environ())
-			// TODO: enable restart
-			//err = syscall.Exec(execFile, os.Args, os.Environ())
-			//if err != nil {
-			//	utils.GetLogInstance().Crit("Failed to restart program after resharding" , "error", err)
-			//}
-			//os.Exit(0)
+			args := os.Args
```
Comment (LeoHChen) [inline node/node_handler.go]: please wrap the argument setting in a function.

File: node/node_handler.go [hunk 9]
```diff
@@ -601,23 +628,43 @@ func (node *Node) processEpochShardState(epochShardState *types.EpochShardState)
...
+				//if arg == newNodeFlag {
+				//	args[i] = ""  // remove new node flag
+				//}
+			}
+			if !hasShardID {
+				args = append(args, shardIDFlag)
+				args = append(args, strconv.Itoa(int(myShard)))
+			}
+			utils.GetLogInstance().Info("Restarting program", "args", args, "env", os.Environ())
+			err = syscall.Exec(execFile, args, os.Environ())
```
Comment (LeoHChen) [inline node/node_handler.go]: do you need to save any state, flush file, save db, before you restart the program?

File: node/node_handler.go [hunk 10]
```diff
@@ -601,23 +628,43 @@ func (node *Node) processEpochShardState(epochShardState *types.EpochShardState)
...
+			if err != nil {
+				utils.GetLogInstance().Crit("Failed to restart program after resharding", "error", err)
+			}
 		}
 	} else {
-		utils.GetLogInstance().Info(fmt.Sprintf("[Resharded][epoch:%d]  Somehow I got kicked out", epoch), "BlsPubKey", hex.EncodeToString(myBlsPubKey))
-		//os.Exit(0)
+		utils.GetLogInstance().Info(fmt.Sprintf("[Resharded][epoch:%d]  Somehow I got kicked out. Exiting", epoch), "BlsPubKey", hex.EncodeToString(myBlsPubKey))
+		os.Exit(0)
```
Comment (LeoHChen) [inline node/node_handler.go]: use a different exit value, please.

File: node/node_handler.go [hunk 11]
```diff
@@ -631,3 +678,33 @@ func lookPath() (argv0 string, err error) {
...
 }
+
+func (node *Node) storeEpochShardState(epochShardState *types.EpochShardState) {
+	byteBuffer := bytes.NewBuffer([]byte{})
+	encoder := gob.NewEncoder(byteBuffer)
+	err := encoder.Encode(epochShardState)
+	if err != nil {
+		utils.GetLogInstance().Error("[Resharded] Failed to encode epoch shard state", "error", err)
+	}
+	err = ioutil.WriteFile("./tmp_log/epoch_shard_state"+node.SelfPeer.IP+node.SelfPeer.Port, byteBuffer.Bytes(), 0644)
```
Comment (LeoHChen) [inline node/node_handler.go]: is it okay if the tmp_log directory doesn't exist?

File: node/node_newblock.go [hunk 12]
```diff
@@ -31,7 +31,7 @@ func (node *Node) WaitForConsensusReady(readySignal chan struct{}, stopChan chan
 			// keep waiting for Consensus ready
 			select {
 			case <-readySignal:
-				time.Sleep(100 * time.Millisecond) // Delay a bit so validator is catched up (test-only).
+				time.Sleep(5000 * time.Millisecond) // Delay a bit so validator is catched up (test-only).
```
Comment (LeoHChen) [inline node/node_newblock.go]: 5 seconds? why?

File: core/resharding.go (hunk 2 above)
Comment (harmony-ek) [reply core/resharding.go]: This is probably a leak from testing (I have the same change in my local).  Below please find/accept the revert.

```suggestion
	GenesisShardSize = 50
```

File: node/node_newblock.go (hunk 12 above)
Comment (rlan35) [reply node/node_newblock.go]: This PR is not ready for review yet. I will clean up after it works with the master code.

File: cmd/harmony/main.go (hunk 1 above)
Comment (rlan35) [reply cmd/harmony/main.go]: This is mistakenly committed code. I already cleaned it locally. [PR not ready for review yet]

Comment (rlan35) [reply cmd/harmony/main.go]: Test code, reverted.

File: core/resharding.go (hunk 2 above)
Comment (rlan35) [reply core/resharding.go]: Test code reverted.

File: core/resharding.go (hunk 3 above)
Comment (rlan35) [reply core/resharding.go]: reverted.

File: core/resharding.go (hunk 6 above)
Comment (rlan35) [reply core/resharding.go]: reverted

File: core/resharding.go (hunk 5 above)
Comment (rlan35) [reply core/resharding.go]: Done.

File: node/node.go (hunk 7 above)
Comment (rlan35) [reply node/node.go]: reverted

File: node/node_handler.go (hunk 8 above)
Comment (rlan35) [reply node/node_handler.go]: Done.

File: node/node_handler.go (hunk 9 above)
Comment (rlan35) [reply node/node_handler.go]: stored in storeEpochShardState

File: core/resharding.go (hunk 4 above)
Comment (rlan35) [reply core/resharding.go]: reverted

File: node/node_newblock.go (hunk 12 above)
Comment (rlan35) [reply node/node_newblock.go]: reverted

File: node/node_handler.go (hunk 10 above)
Comment (rlan35) [reply node/node_handler.go]: done

File: node/node_handler.go (hunk 11 above)
Comment (rlan35) [reply node/node_handler.go]: No it doesn't. Changed to store in ./epoch_shard_state

**Pull request merged**